src/gadgets/scalar/range.rs

```diff
@@ -0,0 +1,249 @@
+//! Collection of range-checking gadgets for Bls12_381 scalars.
+use crate::gadgets::GadgetErrors;
+use anyhow::{Error, Result};
+use dusk_plonk::prelude::*;
+
+/// Builds a complex range-proof (not bounded to a pow_of_two) given a
+/// composer, the max range and the witness.
+///
+/// Checks that `witness < max_range` returning a boolean `Variable`.
+pub fn single_complex_range_proof(
+    composer: &mut StandardComposer,
+    witness: BlsScalar,
+    max_range: BlsScalar,
+) -> Result<Variable, Error> {
+    let num_bits = bits_count(max_range);
+    let closest_pow_of_two = BlsScalar::from(2u64).pow(&[num_bits, 0, 0, 0]);
+    // Compute b' max range.
+    let b_prime = closest_pow_of_two - max_range;
+    // Obtain 128-bit representation of `witness + b'`.
+    let bits = scalar_to_bits(&(witness + b_prime));
+
+    // Create 0 as witness value
+    let zero = composer.add_constant_witness(BlsScalar::zero());
+
+    let mut var_accumulator = zero;
+
+    let accumulator =
+        bits[..]
+            .iter()
+            .enumerate()
+            .fold(BlsScalar::zero(), |scalar_accum, (idx, mut bit)| {
+                if idx >= num_bits as usize {
+                    bit = &0u8;
```

Comment:
We are mutating the pointer to the address of `0u8`, shouldn't we deref the pointer and set its contents to `0u8` instead?

`*bit = 0`

Reply:
Makes complete sense.

Reply:
After trying your approach it does not work.
Basically, I assume that since we're iterating over the bits, we get a reference to the closure parameters. 
Therefore, I cannot do `&mut bit` in the closure attributes def nor dereference it.